### PR TITLE
fix: save `.env` files under `node_modules/.generated`

### PR DESCRIPTION
**Pull request by szymonrybczak**
Summary:
---------

Fixes https://github.com/react-native-community/cli/issues/2307

From now we're saving `.packager.env` and `launchPackager.command` under `${ctx.root}/node_modules/.generated`, so that setups like Yarn v4 with `node-linker: pnpm` will also work.

Test Plan:
----------

1. Clone the repository and do all the required steps from the [Contributing guide](https://github.com/react-native-community/cli/blob/main/CONTRIBUTING.md)
2. Run `yarn start` in other project on port 8081.
3. Run this command:

```sh
node /path/to/react-native-cli/packages/cli/build/bin.js run-ios
```
4. and you should see this output:
```
> yarn ios
info Another process is running on port 8081.
✔ Use port 8082 instead? › Yes
info Found Xcode project "CliStartServerYarnPnpm.xcodeproj"
```

Checklist
----------

- [x] Documentation is up to date to reflect these changes.
- [x] Follows commit message convention described in [CONTRIBUTING.md](https://github.com/react-native-community/cli/blob/main/CONTRIBUTING.md#commit-message-convention)

### CODE REVIEW

**Review by TMisiukiewicz — APPROVED** [review]

**Comment by szymonrybczak** [comment]
@tido64 mind looking at this PR? :pray:

File: packages/cli-tools/src/startServerInNewWindow.ts [hunk 1]
```diff
@@ -34,7 +36,22 @@ function startServerInNewWindow(
...
+
+  if (!generatedPath) {
+    const newPath = path.join(projectRoot, 'node_modules', '.generated');
+
+    try {
+      await mkdir(newPath);
+    } catch (e) {
+      throw new CLIError(`Failed to create ${newPath}`);
+    }
+
```
Comment (tido64) [inline packages/cli-tools/src/startServerInNewWindow.ts:51]: I don't think we need `fs-extra` here. Node added `recursive` in v10.12.0:
```suggestion
    fs.mkdirSync(newPath, { recursive: true, mode: 0o755 });
```

**Review by tido64 — APPROVED** [review]
Left a comment. Everything else looks fine to me.

File: packages/cli-tools/src/startServerInNewWindow.ts [hunk 2]
```diff
@@ -1,15 +1,15 @@
...
 import logger from './logger';
 import chalk from 'chalk';
+import {findPackageDependencyDir} from './findPackageDependencyDir';
 
 const ERROR = `a dev server manually by running ${chalk.bold(
   'npm start',
 )} or ${chalk.bold('yarn start')} in other terminal window.`;
 
-function startServerInNewWindow(
+async function startServerInNewWindow(
```
Comment (Rexogamer) [inline packages/cli-tools/src/startServerInNewWindow.ts:12]: this doesn't appear to be necessary after the other review comment:
```suggestion
function startServerInNewWindow(
```

Comment (szymonrybczak) [reply packages/cli-tools/src/startServerInNewWindow.ts]: yeah, good catch 👍

**Pull request merged**